### PR TITLE
move the etcd config observer to library-go

### PR DESCRIPTION
**Pull request by p0lyn0mial**
this PR moves existing observer to the shared library as it will be used by CAO and OAS-O.

I slightly changed the code so that it supports passing a custom path under which the configuration will be saved.

### CODE REVIEW

**Comment by p0lyn0mial** [comment]
/hold 

for https://github.com/openshift/cluster-openshift-apiserver-operator/pull/371 and squashing

File: pkg/operator/configobserver/apiserver/listers.go [hunk 1]
```diff
@@ -1 +1,27 @@
...
+// Listers provides a unified interface for config observers used in apiserver package
+type Listers interface {
+	APIServerLister
+	ConfigmapLister
+	EndpointsLister
+}
+
+type APIServerLister interface {
+	APIServerLister() configlistersv1.APIServerLister
+	PreRunHasSynced() []cache.InformerSynced
```
Comment (sttts) [inline pkg/operator/configobserver/apiserver/listers.go:10]: what is this? Why only in one?

File: pkg/operator/configobserver/apiserver/observe_etcd_endpoints.go [hunk 2]
```diff
@@ -104,3 +104,77 @@ func ObserveStorageURLs(genericListers configobserver.Listers, recorder events.R
 
 	return observedConfig, errs
 }
+
+// observeStorageURLsFromOldEndPoint observes the storage URL config.
+func observeStorageURLsFromOldEndPoint(genericListers configobserver.Listers, recorder events.Recorder, currentConfig map[string]interface{}) (map[string]interface{}, []error) {
```
Comment (sttts) [inline pkg/operator/configobserver/apiserver/observe_etcd_endpoints.go]: this long func surprises me in an "adjust" commit.

File: pkg/operator/configobserver/apiserver/listers.go [hunk 3]
```diff
@@ -2,27 +2,15 @@ package apiserver
 
 import (
 	configlistersv1 "github.com/openshift/client-go/config/listers/config/v1"
-	corelistersv1 "k8s.io/client-go/listers/core/v1"
```
Comment (sttts) [inline pkg/operator/configobserver/apiserver/listers.go]: could this commit be factored into the previous ones?

File: pkg/operator/configobserver/etcd/listers.go [hunk 4]
```diff
@@ -1,4 +1,4 @@
-package etcdobserver
```
Comment (sttts) [inline pkg/operator/configobserver/etcd/listers.go]: factor into previous commit

**Comment by hexfusion** [comment]
etcd end of this looks good, more or less the same logic as used by OAS. The only comment is it would be nice to have a single version of this defined I think the only difference between KAS and OAS here is KAS appends localhost to etcdURLs.

**Comment by p0lyn0mial** [comment]
> etcd end of this looks good, more or less the same logic as used by OAS

@hexfusion it must look exactly the same as the logic used by OAS :) The only change I did was to add support for passing a custom path under which the configuration will be saved.

**Comment by p0lyn0mial** [comment]
> The only comment is it would be nice to have a single version of this defined I think the only difference between KAS and OAS here is KAS appends localhost to etcdURLs

let me see what it takes to change the code to support KAS.

**Comment by p0lyn0mial** [comment]
it looks like KAS:

1. doesn't support a fall back function when the config map wasn't found https://github.com/openshift/cluster-kube-apiserver-operator/blob/master/pkg/operator/configobservation/etcdendpoints/observe_etcd_endpoints.go#L54

2. always appends 'localhost' https://github.com/openshift/cluster-kube-apiserver-operator/blob/master/pkg/operator/configobservation/etcdendpoints/observe_etcd_endpoints.go#L91

3. supports having two config paths (old and new) - https://github.com/openshift/cluster-kube-apiserver-operator/blob/master/pkg/operator/configobservation/etcdendpoints/observe_etcd_endpoints.go#L26

**Comment by p0lyn0mial** [comment]
the proof PRs look good https://github.com/openshift/cluster-kube-apiserver-operator/pull/912 and https://github.com/openshift/cluster-openshift-apiserver-operator/pull/371

**Comment by hexfusion** [comment]
thanks @p0lyn0mial 

/lgtm

**Pull request merged**